Use regex-like symbols to get 'latest' and 'base' versions

lightandluck opened the pull request:
Resolves #180 

I went with the regex-like operators because I like symbols more than words.

I'm probably missing something about checking for html/url encoding but I'm not sure what to look out for. This seemed too straight-forward but is working.

Should I use a regex to match instead of [equality operators](https://github.com/edgi-govdata-archiving/web-monitoring-ui/blob/b72f4d5b80d41121848c6c6aca2b42215cb59c68/src/components/page-details.jsx#L180-L181)?

danielballan commented:
Just weighing in the naming: Regex symbols seem fine.  Eventually we might additionally support more special symbols (probably words) as discussed in https://github.com/edgi-govdata-archiving/web-monitoring-ui/issues/173.

In current master, I believe all URLs serve as permalinks, which is nice. Will this break that? Obviously the meaning of `$` changes as versions get added, and `^` could change occasionally two, if we ingest older sources of data (mainly thinking of IA). Would it be better to make `$` and `^` redirect to the UUID, thus generating a permalink?

Mr0grog commented:
> In current master, I believe all URLs serve as permalinks, which is nice… Would it be better to make $ and ^ redirect to the UUID, thus generating a permalink?

I definitely agree with this! @lightandluck if you did this work up around [line 155](https://github.com/edgi-govdata-archiving/web-monitoring-ui/blob/master/src/components/page-details.jsx#L154-L164) where we handle non-specific/invalid version ranges instead of down in `_versionsToRender()`, I think it would solve that.

lightandluck commented:
So, I think I understood the concern here and made the operators re-direct to Uuid paths now. I actually moved all of the invalid range logic into `_versionsToRender` as well. There were too many permutations and things to check for by keeping them separate, that it made more sense to me to do it this way. I added another variable to track whether to re-direct or not. I definitely need it to be double-checked.

The difficulty I had was that before we would default whenever anything was off with either `from` or `to`, but with these new operators the url is always "malformed" so I went with what seemed most reasonable without a ton of conditionals.

Also, currently we have this shortcut `page/pageId/..Uuid` to go to the previous version. The opposite path `page/pageId/Uuid..` right now is an error and goes to the default latest diff. With the code changes, this began to re-direct instead to `Uuid` and `latest` which didn't pair well. I added a case where that path goes to the next version instead.

A side-effect of all this is that a path will only re-direct to the default latest diff if both `from` and `to` are incorrect Uuids. If either are correct they will follow the behavior above. I'm not sure how to get the default behavior back without the logic getting overly complicated, or if that behavior is absolutely desired.

lightandluck commented:
I'll update the comments I removed once everything gets hammered out, but most of the cases didn't apply anymore.

Mr0grog commented on src/components/page-details.jsx:
```diff
@@ -176,15 +170,34 @@ export default class PageDetails extends React.Component {
     const [fromId, toId] = (this.props.match.params.change || '').split('..');
-    let from = this.state.page.versions.find(v => v.uuid === fromId);
-    let to = this.state.page.versions.find(v => v.uuid === toId);
+    const versions = this.state.page.versions;
```
Now that we have the concept, I think the right default change is `..$`, so:

```js
const [fromId, toId] = (this.props.match.params.change || '..$').split('..');
```

That might help reduce complexity slightly.

Mr0grog commented on src/components/page-details.jsx:
```diff
@@ -176,15 +170,34 @@ export default class PageDetails extends React.Component {
+      }
+      else {
+        from = from || versions.find(v => v.capture_time < to.capture_time) || to;
+      }
```
I think moving all the lookup & decisions about redirects, etc. here makes sense. 👍 

This definitely works. If you want to go with this, great. 👌  That said, I do think this logic is a little hard to follow (and your concern about being able to explain it is maybe also a good indicator of that). It might be clearer to break it down more explicitly, e.g.

```js
// Look up explicit version IDs
if (toId === '$') {
  to = versions[0];
  shouldRedirect = true;
}
else if (toId) {
  to = versions.find(v => v.uuid === toId);
  if (!to) {
    throw new Error(`There is no version “${toId}”`);
  }
}

// ...repeat for fromId

// Handle empty string version IDs (they are relative to the other version)
if (!from && !to) {
  throw new Error('No versions specified');
}
else if (from && !to) {
  to = versions[versions.indexOf(from) - 1] || from;
  shouldRedirect = true;
}
else {
  from = versions[versions.indexOf(to) + 1] || to;
  shouldRedirect = true;
}
```

Or even abstract out the lookup part:

```js
function findVersion (id, versions) {
  if (id === '^') {
    // at the end, check id vs. the object’s uuid to determine whether to redirect:
    //  shouldRedirect = from.uuid === fromId && to.uuid === toId
    return versions[versions.length - 1];
    // or you could use some other structure to indicate this is a “pointer”
    //   {uuid: versions[versions.length - 1].uuid, redirect: true};
    //   {redirect: versions[versions.length - 1]}
  }
  if (id === '$') {
    return versions[0];
  }

  const version = versions.find(v => v.uuid === id);
  if (version) {
    return version;
  }

  throw new Error(`There is no version “${id}”`);
}

let to = toId && findVersion(toId, versions);
let from = fromId && findVersion(fromId, versions);
// now handle relative IDs
```

lightandluck replied:
I tried both methods, but I couldn't get them to cover all the different cases without becoming very complicated or not working right for me. 

For the first suggestion if we throw an error when `if (!to)`, then we wouldn't be able to check for relative paths later with `if (!from && !to)`. I also found abstracting out the lookup a little harder for me to follow.

I was able to simplify the `else` clauses I was using into single-line, consistent short-circuit evaluations and added comments. Hopefully that makes the logic easier to follow. 

Everything still works, so I'll go ahead and merge.

Mr0grog replied:
> For the first suggestion if we throw an error when `if (!to)`, then we wouldn't be able to check for relative paths later with if `(!from && !to)`.

We would, because that error is only thrown if `toId` was specified. i.e. if you specify an ID that doesn’t work out to an actual version, we shouldn’t redirect to some other version; we should tell you that version doesn’t exist. The point was that we should *not* move on to choosing a new version in that case.

lightandluck replied:
I think I understand what you're saying now and must have done something wrong before, but will take a look at this again in the next couple days.

TBH, every variation makes my brain hurt a bit. Is this a normal thing to do with shortcuts in routing? 😵
